daemon/tests/test.rs

```diff
@@ -0,0 +1,396 @@
+    check_process_name(daemon_pid, "test-daemon")?;
+
+    // Kill the daemon and wait for it and the monitor process to die.
+    send_signal(daemon_pid, libc::SIGTERM)?;
+    wait_for_process_to_die(daemon_pid)?;
+    wait_for_child_to_die(&mut child)??;
+    check_file_does_not_exist(&pidfile_name)?;
+
+    Ok(())
+}
+
+#[test]
+fn test_detach() -> Result<()> {
+    // Start the daemon and make sure that the pidfile exists immediately.
+    // We don't wait for the pidfile to get created because the daemon is
+    // supposed to do so before the parent exits.
+    let pidfile_name = pidfile_name()?;
+    let mut child = test_daemon_command()?.arg("--pidfile").arg(&pidfile_name).arg("--detach")
+        .spawn()?;
+    let child_pid = child.id() as libc::pid_t;
+    wait_for_child_to_die(&mut child)??;
+    let daemon_pid = read_pidfile(&pidfile_name)?;
+
+    // Read the pidfile and ensure that it:
+    // - Identifies a real process...
+    // - ...that is not the child process.
+    // - and whose parent is not the child process (rather, it should be 'init', either global
+    //   or in a container).
+    process_exists(daemon_pid)?;
+    assert_ne!(child_pid, daemon_pid);
+    assert_ne!(child_pid, parent_pid(daemon_pid)?);
+
+    // Kill the daemon and wait for it to die, and verify that the pidfile was deleted.
+    send_signal(daemon_pid, libc::SIGTERM)?;
+    wait_for_process_to_die(daemon_pid)?;
+    check_file_does_not_exist(&pidfile_name)?;
+
+    Ok(())
+}
+
+/// Waits for `path` to become a pidfile with a pid other than `old_pid`.  Returns the new pid.
+fn wait_for_pidfile_to_change<P>(path: P, old_pid: libc::pid_t) -> Result<libc::pid_t>
+    where P: AsRef<Path>
+{
+    let new_pid = wait_until(|| match read_pidfile(path.as_ref()) {
+        Ok(new_pid) if new_pid != old_pid => Complete(new_pid),
+        _ => Incomplete
+    })?;
+    Ok(new_pid)
+}
+
+#[test]
+fn test_detach_monitor() -> Result<()> {
+    // Start the daemon and make sure that the pidfile exists immediately.
+    // We don't wait for the pidfile to get created because the daemon is
+    // supposed to do so before the parent exits.
+    let pidfile_name = pidfile_name()?;
+    let mut child = test_daemon_command()?.arg("--pidfile").arg(&pidfile_name).arg("--detach")
+        .arg("--monitor").spawn()?;
+    let child_pid = child.id() as libc::pid_t;
+    wait_for_child_to_die(&mut child)??;
+    let daemon_pid = read_pidfile(&pidfile_name)?;
+
+    // Check process naming and ancestry.
+    check_process_name(daemon_pid, "test-daemon")?;
+    let monitor_pid = parent_pid(daemon_pid)?;
+    assert_ne!(monitor_pid, daemon_pid);
+    assert_ne!(monitor_pid, child_pid);
+    assert_ne!(parent_pid(monitor_pid)?, child_pid);
+
+    // Kill the daemon process, making it look like an abort(),
+    // and wait for a new daemon process to get spawned.
+    //
+    // (Rust has weird behavior for SIGSEGV, so don't use that.)
```

Comment:
how can the language influence the behavior of a signal? or you mean, on receiving sigsegv the rust runtime does something strange?

Reply:
Rust handles SIGSEGV in such a way that you have to kill a process twice to make it die. See https://users.rust-lang.org/t/is-sigsegv-handled-by-rust-runtime/45680.

I'll add a comment.

Arguably this is a Rust bug, I posted a possible fix to Twitter a few days ago: https://twitter.com/Ben_Pfaff/status/1567531418955436034